src/csl/quote.rs

```diff
@@ -0,0 +1,171 @@
+        self.kinds &= (1 << self.depth) - 1;
+    }
+}
+
+impl Default for SmartQuoter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Whether the character is an opening bracket, parenthesis, or brace.
+fn is_opening_bracket(c: char) -> bool {
+    matches!(c, '(' | '{' | '[')
+}
+
+/// Whether a character is interpreted as a newline by Typst.
+#[inline]
+pub fn is_newline(character: char) -> bool {
+    matches!(
+        character,
+        // Line Feed, Vertical Tab, Form Feed, Carriage Return.
+        '\n' | '\x0B' | '\x0C' | '\r' |
+        // Next Line, Line Separator, Paragraph Separator.
+        '\u{0085}' | '\u{2028}' | '\u{2029}'
+    )
+}
+
+/// Decides which quotes to substitute smart quotes with.
+pub struct SmartQuotes<'s> {
+    /// The opening single quote.
+    pub single_open: &'s str,
+    /// The closing single quote.
+    pub single_close: &'s str,
+    /// The opening double quote.
+    pub double_open: &'s str,
+    /// The closing double quote.
+    pub double_close: &'s str,
+}
+
+impl<'s> SmartQuotes<'s> {
+    /// Create a new `Quotes` struct with the given quotes, optionally falling
+    /// back to the defaults for a language and region.
+    ///
+    /// The language should be specified as an all-lowercase ISO 639-1 code, the
+    /// region as an all-uppercase ISO 3166-alpha2 code.
+    ///
+    /// Currently, the supported languages are: English, Czech, Danish, German,
+    /// Swiss / Liechtensteinian German, Estonian, Icelandic, Italian, Latin,
+    /// Lithuanian, Latvian, Slovak, Slovenian, Spanish, Bosnian, Finnish,
+    /// Swedish, French, Swiss French, Hungarian, Polish, Romanian, Japanese,
+    /// Traditional Chinese, Russian, Norwegian, Hebrew and Croatian.
+    ///
+    /// For unknown languages, the English quotes are used as fallback.
+    pub fn get<T: EntryLike>(ctx: &'s Context<'s, T>) -> Self {
+        let default = ("'", "\"");
+
+        Self {
+            single_open: ctx
+                .term(OtherTerm::OpenInnerQuote.into(), TermForm::default(), false)
+                .unwrap_or(default.0),
+            single_close: ctx
+                .term(OtherTerm::CloseInnerQuote.into(), TermForm::default(), false)
+                .unwrap_or(default.0),
+            double_open: ctx
+                .term(OtherTerm::OpenQuote.into(), TermForm::default(), false)
+                .unwrap_or(default.1),
+            double_close: ctx
+                .term(OtherTerm::CloseQuote.into(), TermForm::default(), false)
+                .unwrap_or(default.1),
+        }
+    }
+
+    /// The opening quote.
+    pub fn open(&self, double: bool) -> &'s str {
+        if double {
+            self.double_open
+        } else {
+            self.single_open
+        }
+    }
+
+    /// The closing quote.
+    pub fn close(&self, double: bool) -> &'s str {
+        if double {
+            self.double_close
+        } else {
+            self.single_close
+        }
+    }
+}
```

Comment:
We could have a `#[test]` here to test the Rust code independently from CSL too, maybe with a few trickier cases.